src/Microsoft.Framework.OptionsModel/ConfigurationBinder.cs

```diff
@@ -0,0 +1,290 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Framework.ConfigurationModel;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.Framework.OptionsModel
+{
+    public class ConfigurationBinder
+    {
+        private readonly Stack<IConfiguration> _context;
+        private readonly IConfiguration _startConfiguration;
+
+        private IConfiguration _currentConfigurationContext;
+
+        public ConfigurationBinder([NotNull] IConfiguration configuration)
+        {
+            _context = new Stack<IConfiguration>();
+            _startConfiguration = configuration;
+        }
+
+        public static void Bind(IConfiguration configuration, object merge)
+        {
+            new ConfigurationBinder(configuration).Read(merge);
+        }
+
+        public void Read(object merge)
+        {
+            if (merge == null)
+            {
+                return;
+            }
+
+            _context.Clear();
+            EnterContext(_startConfiguration);
+            VisitObject(merge);
+            ExitContext();
+        }
+
+        public TOptions Read<TOptions>() where TOptions : new()
+        {
+            var options = new TOptions();
+            Read(options);
+            return options;
+        }
+
+        private void VisitObject(object obj)
```

Comment:
I don't think Visit really gives much meaning, would Bind be a better prefix for all of the Visit methods if that's what  each method is doing?  Something like BindObject, BindProperty, BindDictionary, etc...